File: chatServer/app/services/goalService.py
"""
Goal Service
All database operations for the goals collection.
Virtual field calculation is centralised in _enrich() so it is never duplicated.
"""
import logging
from datetime import datetime, timedelta
from typing import Any, Dict, List, Optional

from bson import ObjectId
from motor.motor_asyncio import AsyncIOMotorDatabase

logger = logging.getLogger(__name__)


def _enrich(doc: Dict[str, Any]) -> Dict[str, Any]:
    """
    Convert ObjectIds to strings and attach calculated virtual fields.
    Called once per document so there is no copy-paste across methods.
    """
    doc["_id"] = str(doc["_id"])
    doc["userId"] = str(doc["userId"])

    target = doc.get("targetAmount", 0)
    saved = doc.get("savedAmount", 0)

    doc["progressPercentage"] = round((saved / target) * 100) if target > 0 else 0
    doc["remainingAmount"] = max(0.0, target - saved)

    target_date: Optional[datetime] = doc.get("targetDate")
    if target_date:
        days_remaining = (target_date - datetime.now()).days
        doc["daysRemaining"] = days_remaining
        doc["isOverdue"] = days_remaining < 0 and doc.get("status") != "Completed"
    else:
        doc["daysRemaining"] = 0
        doc["isOverdue"] = False

    return doc


class GoalService:
    def __init__(self, db: AsyncIOMotorDatabase) -> None:
        self.collection = db.goals

    # ------------------------------------------------------------------
    # Read helpers
    # ------------------------------------------------------------------

    async def get_goals_by_user(
        self,
        user_id: str,
        status: Optional[str] = None,
    ) -> List[Dict[str, Any]]:
        """Return all goals for a user, optionally filtered by status."""
        try:
            query: Dict[str, Any] = {"userId": ObjectId(user_id)}
            if status:
                query["status"] = status

            cursor = self.collection.find(query).sort("targetDate", 1)
            return [_enrich(doc) async for doc in cursor]

        except Exception as e:
            logger.error("Error getting goals for user %s: %s", user_id, e)
            raise

    async def get_goal_by_id(
        self,
        goal_id: str,
        user_id: str,
    ) -> Optional[Dict[str, Any]]:
        """Return a single goal owned by the given user."""
        try:
            doc = await self.collection.find_one(
                {"_id": ObjectId(goal_id), "userId": ObjectId(user_id)}
            )
            return _enrich(doc) if doc else None

        except Exception as e:
            logger.error("Error getting goal %s: %s", goal_id, e)
            raise

    async def get_goal_summary(self, user_id: str) -> Dict[str, Any]:
        """Aggregate summary statistics for a user's goals."""
        try:
            # Single pass through the cursor — no second query needed
            goals = await self.get_goals_by_user(user_id)

            total_target = sum(g.get("targetAmount", 0) for g in goals)
            total_saved = sum(g.get("savedAmount", 0) for g in goals)

            return {
                "totalGoals": len(goals),
                "activeGoals": sum(1 for g in goals if g.get("status") == "Active"),
                "completedGoals": sum(1 for g in goals if g.get("status") == "Completed"),
                "pausedGoals": sum(1 for g in goals if g.get("status") == "Paused"),
                "totalTargetAmount": round(total_target, 2),
                "totalSavedAmount": round(total_saved, 2),
                "overallProgress": round((total_saved / total_target) * 100, 2)
                if total_target > 0
                else 0,
            }

        except Exception as e:
            logger.error("Error getting goal summary for user %s: %s", user_id, e)
            raise

    async def get_goals_by_category(
        self, user_id: str
    ) -> Dict[str, List[Dict[str, Any]]]:
        """Return goals keyed by category."""
        try:
            goals = await self.get_goals_by_user(user_id)
            result: Dict[str, List[Dict[str, Any]]] = {}
            for goal in goals:
                result.setdefault(goal.get("category", "Other"), []).append(goal)
            return result

        except Exception as e:
            logger.error("Error grouping goals by category for user %s: %s", user_id, e)
            raise

    async def get_goals_by_priority(
        self, user_id: str
    ) -> Dict[str, List[Dict[str, Any]]]:
        """Return goals keyed by priority (High / Medium / Low)."""
        try:
            goals = await self.get_goals_by_user(user_id)
            result: Dict[str, List[Dict[str, Any]]] = {"High": [], "Medium": [], "Low": []}
            for goal in goals:
                result[goal.get("priority", "Medium")].append(goal)
            return result

        except Exception as e:
            logger.error("Error grouping goals by priority for user %s: %s", user_id, e)
            raise

    async def get_overdue_goals(self, user_id: str) -> List[Dict[str, Any]]:
        """Return goals whose targetDate has passed and are not Completed."""
        try:
            query = {
                "userId": ObjectId(user_id),
                "targetDate": {"$lt": datetime.now()},
                "status": {"$ne": "Completed"},
            }
            cursor = self.collection.find(query).sort("targetDate", 1)
            return [_enrich(doc) async for doc in cursor]

        except Exception as e:
            logger.error("Error getting overdue goals for user %s: %s", user_id, e)
            raise

    async def get_upcoming_goals(
        self, user_id: str, days: int = 30
    ) -> List[Dict[str, Any]]:
        """Return Active goals due within the next *days* days."""
        try:
            now = datetime.now()
            query = {
                "userId": ObjectId(user_id),
                "targetDate": {"$gte": now, "$lte": now + timedelta(days=days)},
                "status": "Active",
            }
            cursor = self.collection.find(query).sort("targetDate", 1)
            return [_enrich(doc) async for doc in cursor]

        except Exception as e:
            logger.error("Error getting upcoming goals for user %s: %s", user_id, e)
            raise

File: server/services/goalService.js
// services/goalService.js
const Goal = require('../models/goalModel');
const mongoose = require('mongoose');

class GoalService {
  // Create a new goal
  async createGoal(goalData) {
    try {
      const goal = new Goal(goalData);
      await goal.save();
      return goal;
    } catch (error) {
      throw new Error(`Failed to create goal: ${error.message}`);
    }
  }

  // Get goals with filtering and sorting
  async getGoals(userId, filters = {}) {
    try {
      const {
        status,
        category,
        priority,
        sortBy = 'targetDate',
        sortOrder = 'asc',
        limit,
        skip
      } = filters;

      const query = { userId };

      // Apply filters
      if (status) query.status = status;
      if (category) query.category = category;
      if (priority) query.priority = priority;

      // Build sort object
      const sort = {};
      sort[sortBy] = sortOrder === 'desc' ? -1 : 1;

      let goalQuery = Goal.find(query).sort(sort);

      // Apply pagination
      if (skip) goalQuery = goalQuery.skip(parseInt(skip));
      if (limit) goalQuery = goalQuery.limit(parseInt(limit));

      const goals = await goalQuery.exec();
      return goals;
    } catch (error) {
      throw new Error(`Failed to fetch goals: ${error.message}`);
    }
  }

  // Get goal by ID
  async getGoalById(goalId, userId) {
    try {
      const goal = await Goal.findOne({ _id: goalId, userId });
      return goal;
    } catch (error) {
      throw new Error(`Failed to fetch goal: ${error.message}`);
    }
  }

  // Update goal
  async updateGoal(goalId, userId, updateData) {
    try {
      const goal = await Goal.findOneAndUpdate(
        { _id: goalId, userId },
        updateData,
        { new: true, runValidators: true }
      );
      return goal;
    } catch (error) {
      throw new Error(`Failed to update goal: ${error.message}`);
    }
  }

  // Delete goal
  async deleteGoal(goalId, userId) {
    console.log('goalService deleteGoal:', goalId, userId);
    try {
      const result = await Goal.findOneAndDelete({ _id: goalId, userId });
      console.log('goalService result', result);
      return result;
    } catch (error) {
      throw new Error(`Failed to delete goal: ${error.message}`);
    }
  }

  // Add contribution to goal
  async addContribution(goalId, userId, amount) {
    try {
      const goal = await Goal.findOne({ _id: goalId, userId });
      
      if (!goal) {
        return null;
      }

      await goal.updateSavedAmount(amount);
      return goal;
    } catch (error) {
      throw new Error(`Failed to add contribution: ${error.message}`);
    }
  }

  // Mark goal as complete
  async markGoalComplete(goalId, userId) {
    try {
      const goal = await Goal.findOneAndUpdate(
        { _id: goalId, userId },
        { 
          status: 'Completed',
          completedDate: new Date()
        },
        { new: true }
      );
      return goal;
    } catch (error) {
      throw new Error(`Failed to mark goal complete: ${error.message}`);
    }
  }

  // Pause goal
  async pauseGoal(goalId, userId) {
    try {
      const goal = await Goal.findOneAndUpdate(
        { _id: goalId, userId },
        { status: 'Paused' },
        { new: true }
      );
      return goal;
    } catch (error) {
      throw new Error(`Failed to pause goal: ${error.message}`);
    }
  }

  // Resume goal
  async resumeGoal(goalId, userId) {
    try {
      const goal = await Goal.findOneAndUpdate(
        { _id: goalId, userId },
        { status: 'Active' },
        { new: true }
      );
      return goal;
    } catch (error) {
      throw new Error(`Failed to resume goal: ${error.message}`);
    }
  }

  // Get goals by category
  async getGoalsByCategory(userId, category) {
    try {
      const goals = await Goal.find({ userId, category })
        .sort({ targetDate: 1 });
      return goals;
    } catch (error) {
      throw new Error(`Failed to fetch goals by category: ${error.message}`);
    }
  }

  // Get goals by priority
  async getGoalsByPriority(userId, priority) {
    try {
      const goals = await Goal.find({ userId, priority })
        .sort({ targetDate: 1 });
      return goals;
    } catch (error) {
      throw new Error(`Failed to fetch goals by priority: ${error.message}`);
    }
  }

  // Get overdue goals
  async getOverdueGoals(userId) {
    try {
      const today = new Date();
      const goals = await Goal.find({
        userId,
        targetDate: { $lt: today },
        status: { $ne: 'Completed' }
      }).sort({ targetDate: 1 });
      return goals;
    } catch (error) {
      throw new Error(`Failed to fetch overdue goals: ${error.message}`);
    }
  }

  // Get dashboard stats
  async getDashboardStats(userId) {
    try {
      const goals = await Goal.find({ userId });
      const today = new Date();

      const stats = {
        totalGoals: goals.length,
        activeGoals: goals.filter(g => g.status === 'Active').length,
        completedGoals: goals.filter(g => g.status === 'Completed').length,
        pausedGoals: goals.filter(g => g.status === 'Paused').length,
        overdueGoals: goals.filter(g => 
          g.targetDate < today && g.status !== 'Completed'
        ).length,
        totalTargetAmount: goals.reduce((sum, g) => sum + g.targetAmount, 0),
        totalSavedAmount: goals.reduce((sum, g) => sum + g.savedAmount, 0),
        averageProgress: goals.length > 0 ? 
          Math.round(goals.reduce((sum, g) => sum + g.progressPercentage, 0) / goals.length) : 0,
        goalsByCategory: this._getGoalsByCategory(goals),
        goalsByPriority: this._getGoalsByPriority(goals),
        recentlyCompleted: goals
          .filter(g => g.status === 'Completed' && g.completedDate)
          .sort((a, b) => b.completedDate - a.completedDate)
          .slice(0, 5),
        upcomingDeadlines: goals
          .filter(g => g.status === 'Active' && g.targetDate > today)
          .sort((a, b) => a.targetDate - b.targetDate)
          .slice(0, 5)
      };

      return stats;
    } catch (error) {
      throw new Error(`Failed to fetch dashboard stats: ${error.message}`);
    }
  }

  // Get goals summary
  async getGoalsSummary(userId) {
    try {
      return await Goal.getSummaryByProfile(userId);
    } catch (error) {
      throw new Error(`Failed to fetch goals summary: ${error.message}`);
    }
  }

  // Bulk delete goals
  async bulkDeleteGoals(goalIds, userId) {
    try {
      // Validate all goal IDs
      const validIds = goalIds.filter(id => mongoose.Types.ObjectId.isValid(id));
      
      if (validIds.length !== goalIds.length) {
        throw new Error('Some goal IDs are invalid');
      }

      const result = await Goal.deleteMany({
        _id: { $in: validIds },
        userId
      });

      return result;
    } catch (error) {
      throw new Error(`Failed to bulk delete goals: ${error.message}`);
    }
  }

  // Helper method to categorize goals by category
  _getGoalsByCategory(goals) {
    const categories = {};
    goals.forEach(goal => {
      if (!categories[goal.category]) {
        categories[goal.category] = {
          count: 0,
          totalTarget: 0,
          totalSaved: 0
        };
      }
      categories[goal.category].count++;
      categories[goal.category].totalTarget += goal.targetAmount;
      categories[goal.category].totalSaved += goal.savedAmount;
    });
    return categories;
  }

  // Helper method to categorize goals by priority
  _getGoalsByPriority(goals) {
    const priorities = {};
    goals.forEach(goal => {
      if (!priorities[goal.priority]) {
        priorities[goal.priority] = {
          count: 0,
          totalTarget: 0,
          totalSaved: 0
        };
      }
      priorities[goal.priority].count++;
      priorities[goal.priority].totalTarget += goal.targetAmount;
      priorities[goal.priority].totalSaved += goal.savedAmount;
    });
    return priorities;
  }

  // Get goals nearing deadline (within specified days)
  async getGoalsNearingDeadline(userId, days = 30) {
    try {
      const futureDate = new Date();
      futureDate.setDate(futureDate.getDate() + days);

      const goals = await Goal.find({
        userId,
        status: 'Active',
        targetDate: {
          $gte: new Date(),
          $lte: futureDate
        }
      }).sort({ targetDate: 1 });

      return goals;
    } catch (error) {
      throw new Error(`Failed to fetch goals nearing deadline: ${error.message}`);
    }
  }

  // Calculate monthly savings needed for goals
  async calculateMonthlySavingsNeeded(userId) {
    try {
      const activeGoals = await Goal.find({
        userId,
        status: 'Active'
      });

      const today = new Date();
      const goalsWithMonthlySavings = activeGoals.map(goal => {
        const remainingAmount = goal.remainingAmount;
        const monthsRemaining = Math.max(1, Math.ceil(goal.daysRemaining / 30));
        const monthlySavingsNeeded = remainingAmount / monthsRemaining;

        return {
          ...goal.toObject(),
          monthlySavingsNeeded: Math.round(monthlySavingsNeeded * 100) / 100,
          monthsRemaining
        };
      });

      const totalMonthlySavingsNeeded = goalsWithMonthlySavings.reduce(
        (sum, goal) => sum + goal.monthlySavingsNeeded, 0
      );

      return {
        goals: goalsWithMonthlySavings,
        totalMonthlySavingsNeeded: Math.round(totalMonthlySavingsNeeded * 100) / 100
      };
    } catch (error) {
      throw new Error(`Failed to calculate monthly savings needed: ${error.message}`);
    }
  }

  // Check if user has already migrated guest data
  async checkGuestMigration(userId) {
    try {
      const existingGoal = await Goal.findOne({
        userId,
        isGuestMigrated: true
      });
      return !!existingGoal;
    } catch (error) {
      console.error('Error checking guest migration:', error);
      return false;
    }
  }

  // Migrate guest data to user account
  async migrateGuestData(userId, goals) {
    try {
      const migratedGoals = [];

      for (const guestGoal of goals) {
        const goal = new Goal({
          userId,
          name: guestGoal.name,
          description: guestGoal.description,
          targetAmount: guestGoal.targetAmount,
          currentAmount: guestGoal.currentAmount || 0,
          category: guestGoal.category,
          priority: guestGoal.priority || 'medium',
          targetDate: guestGoal.targetDate,
          status: guestGoal.status || 'active',
          isGuestMigrated: true,
          migratedAt: new Date()
        });

        await goal.save();
        migratedGoals.push(goal);
      }

      return {
        success: true,
        message: `${migratedGoals.length} goals migrated successfully`,
        data: {
          migratedCount: migratedGoals.length,
          goals: migratedGoals
        }
      };
    } catch (error) {
      console.error('Error migrating guest data:', error);
      throw error;
    }
  }
}

module.exports = new GoalService();